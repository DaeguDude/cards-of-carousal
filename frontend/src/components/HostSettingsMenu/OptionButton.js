import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const propTypes = {
  isActive: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  children: PropTypes.string.isRequired,
};

const StyledOptionButton = styled.button`
  color: #dddddd;

  ${(props) =>
    props.isActive &&
    css`
      color: var(--primary-text-color);
      :hover {
        background-color: var(--accent-background-color);
      }
    `}

  background-color: var(--primary-background-color);
  display: flex;
  flex-direction: column;
  border: none;
  padding: 30px 0 20px 25px;

  text-align: left;
  font-size: 1rem;
  font-weight: 700;

  @media (min-width: 800px) {
    font-size: 1.5rem;
  }

  @media (min-width: 2000px) {
    font-size: 2.5rem;
  }

  @media (min-width: 3500px) {
    font-size: 4.5rem;
  }
`;

function OptionButton({
  isActive: optionButtonIsActive,
  onClick: optionClick,
  children,
}) {
  return (
    <StyledOptionButton
      isActive={optionButtonIsActive}
      onClick={optionButtonIsActive ? optionClick : undefined}
    >
      {children}
    </StyledOptionButton>
  );
}

OptionButton.propTypes = propTypes;

export default OptionButton;
