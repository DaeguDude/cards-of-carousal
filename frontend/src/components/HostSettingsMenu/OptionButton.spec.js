import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import OptionButton from './OptionButton';

describe('OptionButton', () => {
  describe('rendering', () => {
    it('renders a button with the given text', () => {
      render(
        <OptionButton isActive onClick={() => {}}>
          TEST
        </OptionButton>,
      );

      expect(screen.getByRole('button', { name: 'TEST' })).toBeInTheDocument();
    });

    it('matches the expected snapshot when active', () => {
      const tree = renderer
        .create(
          <OptionButton isActive onClick={() => {}}>
            TEST
          </OptionButton>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches the expected snapshot when not active', () => {
      const tree = renderer
        .create(
          <OptionButton isActive={false} onClick={() => {}}>
            TEST
          </OptionButton>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    it('does not call the onClick callback when it is active but has not been clicked', () => {
      const onClick = jest.fn();

      render(
        <OptionButton isActive onClick={onClick}>
          TEST
        </OptionButton>,
      );

      expect(onClick).not.toHaveBeenCalled();
    });

    it('calls the onClick callback when it is active and clicked', () => {
      const onClick = jest.fn();

      render(
        <OptionButton isActive onClick={onClick}>
          TEST
        </OptionButton>,
      );

      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('calls the onClick callback twice when it is active and clicked twice', () => {
      const onClick = jest.fn();

      render(
        <OptionButton isActive onClick={onClick}>
          TEST
        </OptionButton>,
      );

      userEvent.click(screen.getByRole('button'));
      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(2);
    });

    it('does not call the onClick callback when it is not active and clicked', () => {
      const onClick = jest.fn();

      render(
        <OptionButton isActive={false} onClick={onClick}>
          TEST
        </OptionButton>,
      );

      userEvent.click(screen.getByRole('button'));
      expect(onClick).not.toHaveBeenCalled();
    });
  });

  describe('propTypes', () => {
    it('logs an error when attempting to render without the isActive prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(<OptionButton onClick={() => {}}>text</OptionButton>);

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the isActive prop as a non bool', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionButton isActive={() => {}} onClick={() => {}}>
          text
        </OptionButton>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the onClick prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(<OptionButton isActive>text</OptionButton>);

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the onClick prop as a non function', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionButton isActive onClick={false}>
          text
        </OptionButton>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without children', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(<OptionButton isActive onClick={() => {}} />);

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render children as a non string', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionButton isActive onClick={() => {}}>
          {true}
        </OptionButton>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('does not log any error when all of the needed props are provided', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionButton isActive onClick={() => {}}>
          text
        </OptionButton>,
      );

      expect(consoleSpy).not.toHaveBeenCalled();
    });
  });
});
