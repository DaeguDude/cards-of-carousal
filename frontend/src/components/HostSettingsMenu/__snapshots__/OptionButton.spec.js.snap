// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionButton rendering matches the expected snapshot when active 1`] = `
.c0 {
  color: #dddddd;
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: none;
  padding: 30px 0 20px 25px;
  text-align: left;
  font-size: 1.5rem;
  font-weight: 700;
}

.c0:hover {
  background-color: var(--accent-background-color);
}

<button
  className="c0"
  onClick={[Function]}
>
  TEST
</button>
`;

exports[`OptionButton rendering matches the expected snapshot when not active 1`] = `
.c0 {
  color: #dddddd;
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: none;
  padding: 30px 0 20px 25px;
  text-align: left;
  font-size: 1.5rem;
  font-weight: 700;
}

<button
  className="c0"
>
  TEST
</button>
`;
